src/hotspot/share/opto/idealGraphPrinter.cpp

```diff
@@ -628,9 +630,99 @@ void IdealGraphPrinter::visit_node(Node *n, bool edges, VectorSet* temp_set) {
   }
 }
 
-void IdealGraphPrinter::walk_nodes(Node *start, bool edges, VectorSet* temp_set) {
+void IdealGraphPrinter::print_field(const Node* node) {
+  buffer[0] = 0;
+  stringStream ss(buffer, sizeof(buffer) - 1);
+  ciField* field = get_field(node);
+  uint depth = 0;
+  if (field == NULL) {
+    depth++;
+    field = find_source_field_of_array_access(node, depth);
+  }
+
+  if (field != NULL) {
+    // Either direct field access or array access
+    field->print_name_on(&ss);
+    for (uint i = 0; i < depth; i++) {
+      // For arrays: Add [] for each dimension
+      ss.print("[]");
+    }
+    if (node->is_Store()) {
+      print_prop("destination", buffer);
+    } else {
+      print_prop("source", buffer);
+    }
+  }
+}
+
+ciField* IdealGraphPrinter::get_field(const Node* node) {
+  const TypePtr* adr_type = node->adr_type();
+  Compile::AliasType* atp = NULL;
+  if (C->have_alias_type(adr_type)) {
+    atp = C->alias_type(adr_type);
+  }
+  if (atp != NULL) {
+    ciField* field = atp->field();
+    if (field != NULL) {
+      // Found field associated with 'node'.
+      return field;
+    }
+  }
+  return NULL;
+}
 
+// Try to find the field that is associated with a memory node belonging to an array access.
+ciField* IdealGraphPrinter::find_source_field_of_array_access(const Node* node, uint& depth) {
+  if (!node->is_Mem()) {
+    // Not an array access
+    return NULL;
+  }
+
+  do {
+    if (node->adr_type() != NULL && node->adr_type()->isa_aryptr()) {
+      // Only process array accesses. Pattern match to find actual field source access.
+      node = get_load_node(node);
+      if (node != NULL) {
+        ciField* field = get_field(node);
+        if (field != NULL) {
+          return field;
+        }
+        // Could be a multi-dimensional array. Repeat loop.
+        depth++;
+        continue;
+      }
+    }
+    // Not an array access with a field source.
+    break;
+  } while (depth < 256); // Cannot have more than 255 dimensions
+
+  return NULL;
+}
+
+// Pattern match on the inputs of 'node' to find load node for the field access.
+Node* IdealGraphPrinter::get_load_node(const Node* node) {
+  Node* load = NULL;
+  Node* addr = node->as_Mem()->in(MemNode::Address);
+  if (addr != NULL && addr->is_AddP()) {
+    Node* base = addr->as_AddP()->base_node();
+    if (base != NULL) {
+      if (base->Opcode() == Op_CastPP) {
+        // CastPP node could have been removed.
```

Comment:
There could also be a `CheckCastPP` or multiple casts, right? I would suggest to use `Node::uncast`.

Reply:
Good point! I've updated it.